src/common/mpi.jl

```diff
@@ -0,0 +1,49 @@
+# Convenience functions for working with MPI
+using MPI
+
+"""
+Initialize MPI. Must be called before doing any non-trivial MPI work
+(even in the single-process case). Unlike the MPI.Init() function,
+this can be called multiple times.
+"""
+function mpi_ensure_initialized()
+    # TODO look more closely at interaction between MPI and threads
+    MPI.Initialized() || MPI.Init_thread(MPI.ThreadLevel(3))
+end
+
+"""
+Number of processors used in MPI. Can be called without ensuring initialization.
+"""
+mpi_nprocs() = mpi_nprocs(MPI.COMM_WORLD)
+mpi_nprocs(comm) = (mpi_ensure_initialized(); MPI.Comm_size(comm))
+mpi_master() = (mpi_ensure_initialized(); MPI.Comm_rank(MPI.COMM_WORLD) == 0)
+
+mpi_sum(arr, comm::MPI.Comm) = MPI.Allreduce(arr, +, comm)
+mpi_sum!(arr, comm::MPI.Comm) = MPI.Allreduce!(arr, +, comm)
+mpi_min(arr, comm::MPI.Comm) = MPI.Allreduce(arr, min, comm)
+mpi_min!(arr, comm::MPI.Comm) = MPI.Allreduce!(arr, min, comm)
+mpi_max(arr, comm::MPI.Comm) = MPI.Allreduce(arr, max, comm)
+mpi_max!(arr, comm::MPI.Comm) = MPI.Allreduce!(arr, max, comm)
+mpi_average(arr, comm::MPI.Comm) = mpi_sum(arr, comm) ./ mpi_nprocs(comm)
```

Comment:
That function is really surprising to me, why is this needed? The problem I have is that could be confused with `mpi_mean` as the pendant to `mean`, which would do something else.

Reply:
It's useful eg in the forces test. More generally if we want to make sure some arrays are synced exactly (as opposed to up to machine precision) this is a good way to do it. I don't get your comment: we can call it `mean` if you prefer

Reply:
The nice thing is that `average` is a verb, so `average!` makes sense

Reply:
My problem is that `mpi_sum` / `mpi_max` / `mpi_min` does basically the same thing as `sum`, `max` and `min` as julia functions. There is also `mean` and so why not call this `mpi_mean` to agree with the pattern?

Reply:
We really want to avoid too much similarity between mpi_sum and sum. I thought about overloading sum to avoid mpi_sum, eg something like `sum(comm, arr)`. I decided against it: we want to really separate mpi calls and highlight them as such. So we can do mean if you prefer, but `mean!(arr)` is weird

Reply:
Sure but it exists in `Statistics`. I don't know why they used `mean` instead of `average`, but I prefer to stick to it (and use it here).

Reply:
ok let's go with that